lib/offsite_payments/integrations/paytm.rb

```diff
@@ -0,0 +1,300 @@
+module OffsitePayments #:nodoc:
+  module Integrations #:nodoc:
+    module Paytm
+      require 'openssl'
+      require 'base64'
+      require 'digest'
+      require 'securerandom'
+
+      mattr_accessor :test_url
+      mattr_accessor :production_url
+
+      self.test_url = 'https://pguat.paytm.com/oltp-web/processTransaction'
+      self.production_url = 'https://secure.paytm.in/oltp-web/processTransaction'
+
+      def self.service_url
+        # self.production_url
+        OffsitePayments.mode == :production ? production_url : test_url
+      end
+
+      def self.notification(post, options = {})
+        Notification.new(post, options)
+      end
+
+      def self.return(post, options = {})
+        Return.new(post, options)
+      end
+
+      def self.checksum(params, secret_key, salt = nil)
+        salt = SecureRandom.urlsafe_base64(4 * (3 / 4)) if salt.nil?
+        keys = params.keys
+        str = nil
+        keys = keys.sort
+        keys.each do |k|
+          if str.nil?
+            str = params[k].to_s
+            next
+          end
+          str = str + '|' + params[k].to_s
+        end
+        str = str + '|' + salt
+        check_sum = Digest::SHA256.hexdigest(str)
+        check_sum += salt
+
+        ### encrypting checksum ###
+        aes = OpenSSL::Cipher::Cipher.new('aes-128-cbc')
+        aes.encrypt
+        aes.key = secret_key
+        aes.iv = '@@@@&&&&####$$$$'
+
+        encrypted_data = nil
+        encrypted_data = aes.update(check_sum.to_s) + aes.final
+        encrypted_data = Base64.encode64(encrypted_data)
+
+        check_sum = encrypted_data.delete("\n")
+        # new_pg_encrypt_variable(check_sum, key)
+        check_sum
+      end
+
+      class Helper < OffsitePayments::Helper
+        CHECKSUM_FIELDS = %w(MID ORDER_ID CUST_ID TXN_AMOUNT CHANNEL_ID INDUSTRY_TYPE_ID WEBSITE EMAIL MOBILE_NO).freeze
+
+        mapping :amount, 'TXN_AMOUNT'
+        mapping :account, 'MID'
+        mapping :order, 'ORDER_ID'
+        mapping :description, 'productinfo'
+
+        mapping :customer, first_name: 'firstname',
+                           last_name: 'lastname',
+                           email: 'CUST_ID',
+                           email: 'EMAIL',
```

Comment:
I don't think this will work, you can't have duplicate keys in a hash. Internally, when the `email` key is looked up in the options, only one value will be returned. For example
```
irb(main):001:0> hash = { email: 'foo', email: 'bar' }
(irb):1: warning: duplicated key at line 1 ignored: :email
=> {:email=>"bar"}
irb(main):002:0> hash[:email]
=> "bar"
```